### DISCUSSION TITLE
Transaction Management in Mirth connect

### DISCUSSION THREAD

**Discussion by mkotaprolu** [nextgenhealthcare/connect#6077] [Q&A]
Hi Mirth Experts,

we have destination connector(database writer) that writes into db, it has multiple DML commands, for example
case 1)
if there are update stmt followed by insert stmt.
if the insert fails update stmt should be rolled back and vice versa.

what is the best way to handle it in mirth.
currently we are using Mirth connect 4.4.2

**Comment by rjkroll** [comment]
You should probably wrap each call in a procedure instead, then inside the procedure you do the BEGIN TRAN.....COMMIT/ROLLBACK stuff.  
If the procedure is succesful, you return a value indicating success or not.   Then you use the value to proceed to the next step/procedure call or to exit the transformer step/bypass the next db call.

 If you are trying to do transactions within a transformer step, calling DB Inserts, etc, using Javascript, you are going to have a hard time.

Or, you just make 1 procedure that you call once, that does all your logic within it.

**Reply by pacmano1** [reply]
By "procedure" by a database "stored procedure" or "function" is meant. But I would also be curious on the workflow here - seems to me there should not be a reason to have to roll back anything rather the SQL statements need to be better.

**Reply by mkotaprolu** [reply]
@pacmano1 let's say i want to save data into two different tables(A, B) within the same schema, in the same request, if update/insert  on A succeeds and B gets fails, we need to rollback update/insert on table A, currently if we are using database write destination connector if update/insert  on A succeeds and B fails , A is persisted but not B, but as a transaction management perspective we need to rollback update/insert on table A.

**Reply by mkotaprolu** [reply]
@rjkroll  i was thinking the same, let see what @pacmano1  has in store for us.

**Reply by pacmano1** [reply]
Define "fail" please.  What is the failure condition? If you know the condition you can test for the condition prior to insert. If you're talking just about a general database failure you have a bad database design and it's an entirely different topic.  I would still use stored procedures, of course.

**Reply by mkotaprolu** [reply]
@pacmano1  if we have only read and write privileges of third party database that we want to use, schema changes are not in our control, and not getting communicated to downstream/upstream systems, putting a check can be trivial , but i believe stored procedure is the route to go as you mentioned , thanks for the suggestion.

**Reply by pacmano1** [reply]
If you do not have schema control, then you will not have the ability to create a store procedure.

**Reply by mkotaprolu** [reply]
@pacmano1  yes agreed, but we can write a PL/SQL block within database writer to achieve the same that fits the purpose, kinda similar to procedure, please let me know if it makes sense.

`BEGIN TRANSACTION trans
BEGIN TRY
INSERT INTO vwh_info1 (enterprise_id, practice_id, person_id, created_by, modified_by) VALUES ('3','3','008','3','3')
UPDATE sleep_1 SET hours_sleep = 8  WHERE person_id = '293F';
 IF @@TRANCOUNT > 0
 BEGIN COMMIT TRANSACTION trans
 END
END TRY
BEGIN CATCH
 IF @@TRANCOUNT > 0
 BEGIN ROLLBACK TRANSACTION trans 
 END
END CATCH`

**Reply by pacmano1** [reply]
Why would that insert fail?

**Reply by mkotaprolu** [reply]
assume second insert fail, ignore the technicality of above insert statements, please let me know if PL/SQL block makes sense.

**Reply by pacmano1** [reply]
My point is this:

An insert would typically fail for maybe three reasons: 

1) a constraint on the table (likely), perhaps even your insert should be a `MERGE` or whatever the native upsert SQL is for your db engine.  
2) some native unpredicatable database problem (unlikely)
3) Some table / row lock problem causing the insert to timeout (maybe more frequent than bullet 2)

Bullet 2 and 3 should almost never happen.   Bullet 1 you can test or code for.   So I asked "why does it fail" and you have not answered that question.   

Yes wrapping it a native transaction is best. BUT the SQL writer in SQL mode (vs JS mode) may already do that.

Lastly, your example above should be a single insert statement.

```SQL
INSERT INTO vwh_info (enterprise_id, practice_id, person_id, created_by, modified_by) 
VALUES (
('3','3','008','3','3'),
('4','4','A0F','4','4')
)
```

**Reply by mkotaprolu** [reply]
@pacmano1  i have updated my comment now replaced the second insert with update stmt , can you please let me know how transaction management works here
